functionalization fix for inplace_view ops

bdhirsh opened the pull request:
Another subtle bugfix that came from the LTC <-> Functionalization integration.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75527
* #76358
* #77714
* #77713
* #77132
* #77716
* #77130
* #77129
* #77128
* #77125
* __->__ #77126

bdhirsh committed:
functionalization fix for inplace_view ops

[ghstack-poisoned]

bdhirsh commented on torchgen/gen_functionalization_type.py:
```diff
@@ -395,14 +395,17 @@ def emit_view_functionalization_body(
         {meta_conversion_str}
         reference_tensor_output = at::_ops::{noop_api_name}::call({', '.join(meta_call_args)});
       }}
+      at::functionalization::impl::mutate_view_meta({view_tensor_name}, view_meta);
```
For inplace_view ops, `mutate_view_meta()` is responsible for replaying the newly-added view, and will update the wrapper's size/stride metadata (e.g. with the new strides after running `transpose_()`

Before, I was running `mutate_view_meta()` **before** running the reference meta function, which caused it to run with the input meta tensors already having been mutated. The `reference_tensor_output`'s metadata was then wrong, so in `set_sizes_strides_offset` I'm setting the wrong size/stride info on the wrapper.

ezyang replied:
These PR comments should just be comments in the code, they will be completely lost here.

bdhirsh commented on torchgen/gen_functionalization_type.py:
```diff
@@ -395,14 +395,17 @@ def emit_view_functionalization_body(
       // See  Note [Propagating strides in the functionalization pass]
+      // XLA/LTC don't implement the logic to propagate strides correctly, so we need to rely
+      // on a reference implementation here (instead of relying on the output from the forward lambda
+      // having the correct stride info)
```
My first attempt was just to say "hey we don't actually need to run the meta tensor reference implementation! We can just rely on sizes/strides/storage_offset being correct after running the view op on the inner tensor!

We can't rely on that though because XLA/LTC's implementations for `transpose` won't actually properly compute strides for you.

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

ezyang commented on test/test_functionalization.py:
```diff
@@ -272,7 +273,7 @@ def f(x):
             y = x[0]
             y.add_(tmp)
-            return y
+            return x
```
does the return here matter?

bdhirsh replied:
yep; first I got a repro from LTC tests, then I tried to swizzle this test until I could get a repro from it.

I also added `self.assertEqual(out_ref.size(), out_functional.size())` to `assert_functionalization` above.

This version of the test now fail the new assert above without the change, because the `FunctionalTensorWrapper` for x used to have its sizes set incorrectly.

ezyang commented:
Do the test case modifications actually exercise the change here?

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization fix for inplace_view ops"

Another subtle bugfix that came from the LTC <-> Functionalization integration.




[ghstack-poisoned]
